frame/society/src/lib.rs

```diff
@@ -0,0 +1,1148 @@
+pub enum Judgement {
+	/// The suspension judgement origin takes no direct judgment
+	/// and places the candidate back into the bid pool.
+	Rebid,
+	/// The suspension judgement origin has rejected the candidate's application.
+	Reject,
+	/// The suspension judgement origin approves of the candidate's application.
+	Approve,
+}
+
+/// Details of a payout given as a per-block linear "trickle".
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug, Default)]
+pub struct Payout<Balance, BlockNumber> {
+	/// Total value of the payout.
+	value: Balance,
+	/// Block number at which the payout begins.
+	begin: BlockNumber,
+	/// Total number of blocks over which the payout is spread.
+	duration: BlockNumber,
+	/// Total value paid out so far.
+	paid: Balance,
+}
+
+/// Number of strikes that a member has against them.
+pub type StrikeCount = u32;
+
+/// A vote by a member on a candidate application.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum BidKind<AccountId, Balance> {
+	/// The CandidateDeposit was paid for this bid.
+	Deposit(Balance),
+	/// A member vouched for this bid. The account should be reinstated into `Members` once the
+	/// bid is successful (or if it is rescinded prior to launch).
+	Vouch(AccountId, Balance),
+}
+
+impl<AccountId: PartialEq, Balance> BidKind<AccountId, Balance> {
+	fn check_voucher(&self, v: &AccountId) -> DispatchResult {
+		if let BidKind::Vouch(ref a, _) = self {
+			if a == v {
+				Ok(())
+			} else {
+				Err("incorrect identity")?
+			}
+		} else {
+			Err("not vouched")?
+		}
+	}
+}
+
+// This module's storage items.
+decl_storage! {
+	trait Store for Module<T: Trait<I>, I: Instance=DefaultInstance> as Society {
+		/// The current set of candidates; bidders that are attempting to become members.
+		pub Candidates get(candidates):
+			Vec<(BalanceOf<T, I>, T::AccountId, BidKind<T::AccountId, BalanceOf<T, I>>)>;
+
+		/// The set of suspended candidates.
+		pub SuspendedCandidates get(suspended_candidate):
+			map T::AccountId => Option<(BalanceOf<T, I>, BidKind<T::AccountId, BalanceOf<T, I>>)>;
+
+		/// Amount of our account balance that is specifically for the next round's bid(s).
+		pub Pot get(fn pot) config(): BalanceOf<T, I>;
+
+		/// The most primary from the most recently approved members.
+		pub Head get(head) build(|config: &GenesisConfig<T, I>| config.members.first().cloned()):
+			Option<T::AccountId>;
+
+		/// The current set of members, ordered.
+		pub Members get(fn members) build(|config: &GenesisConfig<T, I>| {
+			let mut m = config.members.clone();
+			m.sort();
+			m
+		}): Vec<T::AccountId>;
+
+		/// The set of suspended members.
+		pub SuspendedMembers get(fn suspended_member): map T::AccountId => Option<()>;
+
+		/// The current bids.
+		Bids: Vec<(BalanceOf<T, I>, T::AccountId, BidKind<T::AccountId, BalanceOf<T, I>>)>;
+
+		/// Members currently vouching, stored ordered
+		Vouching: Vec<T::AccountId>;
```

Comment:
@gavofyork correct me if I am wrong here, but I see no reason for `Vouching` to be a `Vec`.

We only ever add and remove from it a single user at a time.

Reading all the vouchers from storage and doing a binary search to add/remove each time seems way less efficient than simply storing a `map T::AccountId => Option<()>` to keep track of this.

Reply:
that very much remains to be seen, and i suspect empirical testing will be the only way to find out. for now i'm assuming a single DB read of (likely) less than 1KB followed by a binary chop search and insert will be more efficient overall than using the main trie for 30-40 items. i might be wrong though.

Reply:
Since tries are prefixed now by module and storage item, the extra nodes under this branch for a map wont affect performance of any other module or storage item query... right?

Reply:
Since tries are prefixed now by module and storage item, the extra nodes under a branch for a storage map should not affect performance of any other module or storage item queries.

So in this case, it makes sense to use a map.

Will follow up with benchmarks after the holidays.